src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.Unix.cs

```diff
@@ -0,0 +1,156 @@
+
+        internal static void CreateDirectory(string fullPath, UnixFileMode? mode, bool overwriteMetadata, SortedDictionary<string, UnixFileMode>? pendingModes)
+        {
+            // Restrictive mask for creating the missing parent directories while extracting.
+            const UnixFileMode ExtractPermissions = UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute;
+
+            Debug.Assert(!OperatingSystem.IsWindows());
+            Debug.Assert(pendingModes is not null);
+
+            if (Directory.Exists(fullPath))
+            {
+                // Apply permissions to an existing directory when we're overwriting metadata
+                // or the directory was created as a missing parent (stored in pendingModes).
+                if (mode.HasValue)
+                {
+                    bool hasExtractPermissions = (mode & ExtractPermissions) == ExtractPermissions;
+                    if (hasExtractPermissions)
+                    {
+                        bool removed = pendingModes.Remove(fullPath);
+                        if (overwriteMetadata || removed)
+                        {
+                            UnixFileMode umask = UMask;
+                            File.SetUnixFileMode(fullPath, mode.Value & ~umask);
+                        }
+                    }
+                    else
+                    {
+                        if (overwriteMetadata || pendingModes.ContainsKey(fullPath))
+                        {
+                            pendingModes[fullPath] = mode.Value;
+                        }
+                    }
+                }
+                return;
+            }
+
+            if (mode.HasValue)
+            {
+                // Ensure we have sufficient permissions to extract in the directory.
+                if ((mode & ExtractPermissions) != ExtractPermissions)
+                {
+                    pendingModes[fullPath] = mode.Value;
+                    mode = ExtractPermissions;
+                }
+            }
+            else
+            {
+                pendingModes.Add(fullPath, DefaultDirectoryMode);
+                mode = ExtractPermissions;
+            }
+
+            string parentDir = Path.GetDirectoryName(fullPath)!;
+            string rootDir = Path.GetPathRoot(parentDir)!;
+            bool hasMissingParents = false;
+            for (string dir = parentDir; dir != rootDir && !Directory.Exists(dir); dir = Path.GetDirectoryName(dir)!)
+            {
+                pendingModes.Add(dir, DefaultDirectoryMode);
+                hasMissingParents = true;
+            }
+
+            if (hasMissingParents)
+            {
+                Directory.CreateDirectory(parentDir, ExtractPermissions);
+            }
+
+            Directory.CreateDirectory(fullPath, mode.Value);
+        }
+
+        internal static void SetPendingModes(SortedDictionary<string, UnixFileMode>? pendingModes)
```

Comment:
`tar` has this type of handling for directories.

I took the two archives that are created by the `UnixFileMode` and `UnixFileModes_RestrictiveParentDir` tests and gave them to `tar` command for extracting.

This is the output interleaved with syscalls for mkdir and chmod:

```
dir
mkdirat(AT_FDCWD, "dir", 0740)          = 0
file
missing_parent/dir
mkdirat(AT_FDCWD, "missing_parent/dir", 0704) = -1 ENOENT (No such file or directory)
mkdirat(AT_FDCWD, "missing_parent", 0775) = 0
mkdirat(AT_FDCWD, "missing_parent/dir", 0704) = 0
out_of_order_parent/file
mkdirat(AT_FDCWD, "out_of_order_parent", 0775) = 0
out_of_order_parent
mkdirat(AT_FDCWD, "out_of_order_parent", 0701) = -1 EEXIST (File exists)
fchmodat(AT_FDCWD, "out_of_order_parent", 0701) = 0
```

Here you see the permission being applied when we encounter the `out_of_order_parent` later in the archive.

We don't see a restrictive permission being used initially when the `missing_parent` and `out_of_order_parent` are created. It does seem better to do this, since the archive permissions may be less permissive than the default initially.

```
dir
mkdirat(AT_FDCWD, "dir", 0300)          = 0
dir/file
chmod("/proc/self/fd/3", 000)           = 0
```

Here we see the `dir` being created with minimal permissions needed for extracting.
Then later it is set to the correct permission.

Note that in the common case the archives are ordered: parents come before children. In this case `pendingModes` remains empty.